web/frontend/src/components/utils/useConfiguration.tsx

```diff
@@ -0,0 +1,165 @@
+    for (const select of subject.Selects) {
+      let insideOrder = subject.Order.findIndex((id: ID) => id === select.ID) + 1;
+      numberOfQuestion += 1;
+
+      sortedQuestions.push({
+        Order: outerOrder + insideOrder,
+        ParentID: subject.ID,
+        Type: SELECT,
+        Content: select,
+        render: renderSelect,
+      });
+
+      answerList.SelectAnswers.push({
+        ID: select.ID,
+        Answers: new Array<boolean>(select.Choices.length).fill(false),
+      });
+
+      answerList.Errors.push({
+        ID: select.ID,
+        Message: '',
+      });
+    }
+  }
+
+  if (subject.Ranks) {
+    for (const rank of subject.Ranks) {
+      let insideOrder = subject.Order.findIndex((id: ID) => id === rank.ID) + 1;
+      numberOfQuestion += 1;
+
+      sortedQuestions.push({
+        Order: outerOrder + insideOrder,
+        ParentID: subject.ID,
+        Type: RANK,
+        Content: rank,
+        render: renderRank,
+      });
+
+      answerList.RankAnswers.push({
+        ID: rank.ID,
+        Answers: Array.from(Array(rank.Choices.length).keys()),
+      });
+    }
+  }
+
+  if (subject.Texts) {
+    for (const text of subject.Texts) {
+      let insideOrder = subject.Order.findIndex((id: ID) => id === text.ID) + 1;
+      numberOfQuestion += 1;
+
+      sortedQuestions.push({
+        Order: outerOrder + insideOrder,
+        ParentID: subject.ID,
+        Type: TEXT,
+        Content: text,
+        render: renderText,
+      });
+
+      answerList.TextAnswers.push({
+        ID: text.ID,
+        Answers: new Array<string>(text.Choices.length).fill(''),
+      });
+
+      answerList.Errors.push({
+        ID: text.ID,
+        Message: '',
+      });
+    }
+  }
+
+  return numberOfQuestion;
+}
+
+// On Configuration changes, iterate over the top-level Subjects of the Configuration
+// to sort the questions and initialize the corresponding Answers.
+const useConfiguration = (configuration: Configuration) => {
+  const [sortedQuestions, setSortedQuestions] = useState(Array<Question>());
+  const [answers, setAnswers]: [Answers, React.Dispatch<React.SetStateAction<Answers>>] =
+    useState(null);
+
+  useEffect(() => {
+    if (configuration !== null) {
+      let order: number = 0;
+      let questionList = Array<Question>();
+      let answerList: Answers = {
+        SelectAnswers: new Array<SelectAnswer>(),
+        RankAnswers: new Array<RankAnswer>(),
+        TextAnswers: new Array<TextAnswer>(),
+        Errors: new Array<Error>(),
+      };
+
+      for (const subject of configuration.Scaffold) {
+        order += flattenSubject(subject, ROOT_ID, questionList, order, answerList);
+      }
+      questionList.sort((q1, q2) => {
+        return q1.Order - q2.Order;
+      });
+      setSortedQuestions(questionList);
+      setAnswers(answerList);
+    }
+  }, [configuration]);
+
+  return {
+    sortedQuestions,
+    answers,
+    setAnswers,
+  };
+};
+
+export default useConfiguration;
```

Comment:
see my comment on `web/frontend/src/pages/ballot/Show.tsx`. It seems that this can be simplified.